hrl/agent/dsc/classifier/sift_classifier.py

```diff
@@ -0,0 +1,322 @@
+
+    def optimistic_predict(self, state):
+        assert isinstance(state, np.ndarray)
+        return self.optimistic_classifier.predict([state])[0] == 1
+
+    def pessimistic_predict(self, state):
+        assert isinstance(state, np.ndarray)
+        assert state.shape == (84, 84)
+        return self.pessimistic_classifier_for_pos_data.predict([state])[0] == 1 and self.pessimistic_classifier_for_neg_data.predict([state])[0] == -1
+
+    def add_positive_examples(self, images, positions):
+        assert len(images) == len(positions)
+
+        positive_examples = [TrainingExample(img, pos) for img, pos in zip(images, positions)]
+        self.positive_examples.append(positive_examples)
+
+    def add_negative_examples(self, images, positions):
+        assert len(images) == len(positions)
+
+        negative_examples = [TrainingExample(img, pos) for img, pos in zip(images, positions)]
+        self.negative_examples.append(negative_examples)
+
+    @staticmethod
+    def extract_positions(examples):
+        examples = itertools.chain.from_iterable(examples)
+        positions = [example.pos for example in examples]
+        return np.array(positions)
+
+    @staticmethod
+    def construct_image_list(examples):
+        """
+        examples is either self.positive_examples or self.nagetive_examples
+        """
+        images = [example.obs._frames[-1].squeeze() for trajectory in examples for example in trajectory]
+        return images
+
+    def fit_initiation_classifier(self):
+        if len(self.negative_examples) > 0 and len(self.positive_examples) > 0:
+            self.train_two_class_classifier()
+        elif len(self.positive_examples) > 0:
+            self.train_one_class_classifier()
+
+    def train_one_class_classifier(self):
+        # data
+        positive_examples = self.construct_image_list(self.positive_examples)
+        positive_labels = [1 for _ in positive_examples]
+        # pessimistic
+        self.pessimistic_classifier.fit(positive_examples, positive_labels, svm_type='one_class_svm', nu=0.1)
+        # optimistic
+        self.optimistic_classifier.fit(positive_examples, positive_labels, svm_type='one_class_svm', nu=0.01)
+
+    def train_two_class_classifier(self):
+        # data
+        positive_examples = self.construct_image_list(self.positive_examples)
+        positive_labels = [1 for _ in positive_examples]
+        nagative_examples = self.construct_image_list(self.negative_examples)
+        negative_labels = [0 for _ in nagative_examples]
+        X = np.array(positive_examples + nagative_examples)
+        Y = np.array(positive_labels + negative_labels)
+        # optimistic
+        self.optimistic_classifier.fit(X, Y, svm_type='svc')
+        # pessimistic
+        training_predictions = self.optimistic_classifier.predict(X)
+        positive_training_examples = X[training_predictions == 1]
+        negative_training_examples = X[training_predictions == 0]
```

Comment:
Rewrite as `X[training_predictions != 1]` to apply to both oc and tc classifiers.